pkg/agent/controller/service_import_aggregator.go

```diff
@@ -57,28 +60,63 @@ func (a *ServiceImportAggregator) updateOnCreateOrUpdate(name, namespace string)
 				a.clusterID, existing.Name, existing.Status.Clusters)
 		}
 
-		// TODO - merge port info
+		return a.setServicePorts(existing)
 	})
 }
 
+func (a *ServiceImportAggregator) setServicePorts(si *mcsv1a1.ServiceImport) error {
+	// We don't set the port info for headless services.
+	if si.Spec.Type != mcsv1a1.ClusterSetIP {
+		return nil
+	}
+
+	serviceName := si.Annotations[mcsv1a1.LabelServiceName]
+	serviceNamespace := si.Annotations[constants.LabelSourceNamespace]
+
+	list, err := a.brokerClient.Resource(endpointSliceGVR).Namespace(a.brokerNamespace).List(context.Background(), metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			discovery.LabelManagedBy:       constants.LabelValueManagedBy,
+			constants.LabelSourceNamespace: serviceNamespace,
+			mcsv1a1.LabelServiceName:       serviceName,
+		}).String(),
+	})
+	if err != nil {
+		return errors.Wrapf(err, "error listing the EndpointSlices associated with service %s/%s",
+			serviceNamespace, serviceName)
+	}
+
+	si.Spec.Ports = make([]mcsv1a1.ServicePort, 0)
+
+	for i := range list.Items {
+		eps := a.converter.toEndpointSlice(&list.Items[i])
+		si.Spec.Ports = slices.Union(si.Spec.Ports, a.converter.toServicePorts(eps.Ports), func(p mcsv1a1.ServicePort) string {
+			return fmt.Sprintf("%s%s%d", p.Name, p.Protocol, p.Port)
```

Comment:
What about `p.AppProtocol`? It is another field in ServicePort but not seen it used much ouside of service without selectors. We can revisit this if run into any bugs due to services using appProtocol.

Reply:
OK - makes sense. I'll push a separate commit to propagate it.